packages/liveblocks-react-ui/src/components/internal/AiChatAssistantMessage.tsx

```diff
@@ -154,60 +131,41 @@ export const AiChatAssistantMessage = memo(
 );
 
 function AssistantMessageContent({
-  content,
-  chatId,
-  messageId,
+  message,
   components,
 }: {
-  content: AiAssistantContentPart[];
-  chatId: string;
-  messageId: MessageId;
+  message: UiAssistantMessage;
   components: Partial<GlobalComponents> | undefined;
 }) {
+  const content = message.content ?? message.contentSoFar;
+
   // A message is considered to be in "reasoning" state if it only contains reasoning parts and no other parts.
   const isReasoning =
     content.some((part) => part.type === "reasoning") &&
     content.every((part) => part.type === "reasoning");
 
   return (
     <div className="lb-ai-chat-message-content">
-      {content.map((part, index) => {
-        switch (part.type) {
-          case "text": {
-            return (
-              <TextPart
-                key={index}
-                text={part.text}
-                components={components}
-                className="lb-ai-chat-message-text"
-              />
-            );
-          }
-          case "tool-invocation": {
-            return (
-              <ToolInvocationPart
-                key={index}
-                chatId={chatId}
-                messageId={messageId}
-                part={part}
-              />
-            );
-          }
-          case "reasoning": {
-            return (
-              <ReasoningPart
-                key={index}
-                text={part.text}
-                isPending={isReasoning}
-                components={components}
-              />
-            );
-          }
-          default: {
-            return null;
-          }
-        }
-      })}
+      <AiMessage.Content
+        message={message}
+        components={{
+          TextPart: ({ part }) => (
+            <TextPart
+              text={part.text}
+              components={components}
+              className="lb-ai-chat-message-text"
+            />
+          ),
+
+          ReasoningPart: ({ part }) => (
+            <ReasoningPart
+              text={part.text}
+              isPending={isReasoning}
```

Comment:
Could we give `isReasoning` (and in the future things like the time it took/is taking) as a prop?

```
ReasoningPart: ({ part, status }) => status === "reasoning" ? "Reasoning…" : "Reasoned"
```

Reply:
Yeah maybe not `status`, I do think the concept of "pending when it's the only part" could be first-class especially if/when it comes with duration info (`{ part, isPending, duration /* The reasoning total duration in ms, when pending and when done */ }`) but that not a requirement now.

For `isTheOnlyPart`, since this component does the looping, would it make sense to give `{ part, index, content }` similarly to `Array.map` to allow `isTheOnlyPart` and anything else?

Reply:
Oh yes, I like passing `index` (and `duration`!) as well. Duration would require a bit more tracking, but it would indeed be pretty nice to have (for all parts, not just reasoning).

What do you think, @nimeshnayaju ?

Reply:
I can see how `index` makes sense - although if we decide to do grouping, not sure if the idea of index would still make sense. 

`isPending` makes a lot of sense too as it applies to all parts. 

`duration`, on the other hand, is more complex and would most definitely be a value generated on the server. It is also possible for the same assistant message to have interleaving reasoning and text parts (e.g. `thinking`, `text`, `thinking`), so I think this will require more thought.

Reply:
> _"I can see how index makes sense - although if we decide to do grouping, not sure if the idea of index would still make sense."_

Good thought. Maybe we need two indexes? A group index, and an index within the group (starting at 0 in each group).

Reply:
> _"duration, on the other hand, is more complex and would most definitely be a value generated on the server. It is also possible for the same assistant message to have interleaving reasoning and text parts (e.g. thinking, text, thinking), so I think this will require more thought."_

Absolutely, we can come back to this idea later.

Reply:
I've implemented the `isPending` prop, but named it `isStreaming` instead. While at it, I'm also passing down the `index` as a prop, in case the part wants it. See 207402cbd701f0411aa6cd34500f0d13e1fb9e24.

Reply:
I like it 👍 For `lb-ai-chat-pending`, this is a generic class that applies a shimmer effect so it can be used for anything loading/pending (for example we might use it on the title in `AiTool` if the spinner is too aggressive), I think it still makes sense to call it "pending" and not restricting it to "streaming"

Reply:
The boolean should reflect that the part's data can still be expected to change (so indeed useful to implement the shimmer effect). I wouldn't mind going back to pending, but I wanted to kind of avoid it since messages can also be said to be "pending" if they are still in either "generating" or "awaiting-tool" state. And a tool invocation can be "pending" if it's in "receiving" or "executing" status. So my main worry is… is it going to be clear enough what it means for a part to be "pending"? Maybe we should flip the bool and call it `isFinalized`?

Reply:
Sorry I meant for the CSS class specifically, `isStreaming` makes sense on the JS side but this CSS class is used in other contexts so it's not tied to this streaming concept. Hopefully that makes sense, imagine a class named `.loading` which would make any element it's applied to fade in and out, we would use it in many contexts without naming it differently for each.

Reply:
Ohhhh – whoops! My bad for renaming it. I wanted to make it consistent as I thoughtit was only used here. I'll rename the CSS class back then, and keep the rest 👍